docs/design/assignment.md

```diff
@@ -0,0 +1,357 @@
+[bitwise](expressions/bitwise.md) operator `$`, a corresponding compound
+assignment `$=` is provided that performs the operation in-place:
+
+```
+// Same as `a = a + 1;`
+a += 1;
+// Same as `a = a << 3;`
+a <<= 3;
+```
+
+In addition, increment and decrement operators are provided:
+
+```
+// Same as `a = a + 1;`
+++a;
+// Same as `a = a - 1;`
+--a;
+```
+
+These simple assignment, compound assignment, increment, and decrement operators
+can only be used as complete statements, not as subexpressions of other
+operators, even when parenthesized:
+
+```
+var n: i32;
+// Error, assignment is not permitted as a subexpression.
+if (F() and (n = GetValue()) > 5) {
+}
+```
+
+User-defined types can define the meaning of these operations by
+[implementing an interface](#extensibility) provided as part of the Carbon
+standard library.
+
+## Syntax
+
+The operands of these operators can be any [expression](expressions/README.md).
+However, the first operand must be modifiable because it is passed to an
+`[addr self: Self*]` parameter, which disallows most expression forms other
+than:
+
+-   The name of a `var` binding.
+-   A dereference of a pointer.
+-   Array indexing that produces a modifiable result.
+-   Member access naming a field, where the object is one of these expressions.
+
+## Simple assignment semantics
+
+A simple assignment statement is intended to exactly mirror the semantics of
+initialization. The following two code snippets should have the same meaning if
+they are both valid:
+
+```
+// Declare and initialize.
+var v: T = init;
+```
+
+```
+// Declare separately from initialization.
+// Requires that `T` has an unformed state.
+var v: T;
+v = init;
+```
+
+This equivalence is not enforced, but when an object is in an unformed state,
+running the assignment function is _optional_, just like running the destructor
+is. If the assignment function is not run, the object will be directly
+initialized from the right-hand side instead. The type is still required to
+implement `AssignWith` for the assignment to be valid.
+
+```
+class C { ... }
+fn F() -> C {
+  returned var c: C = {...};
+  // `&c` here is `&x` for the first call to `F()`.
+  // `&c` here can be `&y` for the second call  to `F()`.
+  return var;
+}
+fn G() {
+  var x: C = F();
+  var y: C;
+  y = F();
+}
+```
+
+## Compound assignment semantics
+
+The syntax `a $= b;` is intended to be syntactic sugar for `a = a $ b;`, except
+as follows:
+
+-   A type might be able to provide a more efficient implementation for the
+    compound assignment form than for the uncombined form.
+-   A type might not be able to, or might not want to, provide the uncombined
+    form at all, for example because creating a new instance requires additional
+    resources that might not be available, such as a context object or an
+    allocator.
+
+This syntactic sugar is provided by a [default implementation](#defaults) of
+`$=` in terms of `$` and `=`.
```

Comment:
Actually hitting the button to submit my wording tweak....

```suggestion
The syntactic sugar implementation is provided by a [default implementation](#defaults) of
`$=` in terms of `$` and `=`.
```

Reply:
```suggestion
The syntactic sugar is implemented by a [default implementation](#defaults) of
`$=` in terms of `$` and `=`.
```

Reply:
LG